### PR TITLE
Vosk speech recognition transcription service

### PR DESCRIPTION
**Pull request by nshmyrev**
This is an initial attempt to implement speech recognition with the open source[ Vosk transcription server](https://github.com/alphacep/vosk-server). The server supports several languages and runs without internet. Installation is done with a simple docker command.

Everything seems in place, just need to test it properly.

### CODE REVIEW

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        implements TranscriptionService {
+
+    /**
+     * The logger for this class
+     */
+    private final static Logger logger
+            = Logger.getLogger(VoskTranscriptionService.class);
+
+    /**
+     * Property name for username in credential JSON of IBM Watson
```
Comment (nikvaessen) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:60]: IBM watson comment is out of context

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+     * speech-to-text service.
+     */
+    public final static String P_URL
```
Comment (nikvaessen) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:63]: I think the variable name should be a bit more descriptive.

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+            = "org.jitsi.jigasi.transcription.vosk.websocket_url";
+
+    public final static String DEFAULT_P_URL = "ws://localhost:2700";
```
Comment (nikvaessen) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:66]: same

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+    private final static String EOF_MESSAGE = "{\"eof\" : 1}";
+
+    private String websocketUrl;
```
Comment (nikvaessen) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:70]: should this URL ever change? If it does not, can it be final?

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+ * limitations under the License.
+ */
+package org.jitsi.jigasi.transcription;
+
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketClose;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketConnect;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketError;
+import org.eclipse.jetty.websocket.api.annotations.OnWebSocketMessage;
+import org.eclipse.jetty.websocket.client.WebSocketClient;
```
Comment (nikvaessen) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:25]: The Jitsi codebase styleguide specifies to use * imports unless conflicts need to be resolved

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+                        false,
+                        "C",
+                        0,
+                        new TranscriptionAlternative(msg)));
+            }
+        }
+
+        @OnWebSocketError
+        public void onError(Throwable cause) {
+            cause.printStackTrace(System.out);
```
Comment (nikvaessen) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:196]: Should this be System.err instead?

Comment (damencho) [reply src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:196]: logger actually?

**Comment by nshmyrev** [comment]
Thank you for your comments, I'll try to address them. Still have trouble to setup the running transcription with docker. @damencho do I understand correctly that now the recommended way to implement transcription is with MUC/brewery setup, not with component and the current docker setup https://github.com/jitsi/docker-jitsi-meet does not support MUC yet.

**Comment by damencho** [comment]
> the recommended way to implement transcription is with MUC/brewery setup, not with component 

Correct.





> the current docker setup https://github.com/jitsi/docker-jitsi-meet does not support MUC yet.

That is not true. By default docker uses mucs: https://github.com/jitsi/docker-jitsi-meet/blob/master/jigasi/rootfs/defaults/sip-communicator.properties#L62

**Comment by nshmyrev** [comment]
Hm, I still can't get it working even with default setup, without changes I get [the problem about Reading from SSRC while it is not known as a participant](https://community.jitsi.org/t/jigasi-transcription-problem-reading-from-ssrc-while-it-is-not-known-as-a-participant/52412). If I updated to latest master, I get this, transcriber exits a moment after I invite it as `jitsi_meet_transcribe` participant:

```
jigasi_1   | Jigasi 2020-06-30 21:14:00.212 INFO: [93] org.jitsi.jigasi.TranscriptionGatewaySession.log() Transcriber: Media Device Audio
jigasi_1   | Jigasi 2020-06-30 21:14:00.244 INFO: [93] service.protocol.media.MediaHandler.registerDynamicPTsWithStream().1003 Dynamic PT map: 126=rtpmap:-1 telephone-event/8000; 111=rtpmap:-1 opus/48000/2 fmtp:useinbandfec=1;minptime=10; 103=rtpmap:-1 unknown/90000; 
jigasi_1   | Jigasi 2020-06-30 21:14:00.244 INFO: [93] service.protocol.media.MediaHandler.registerDynamicPTsWithStream().1020 PT overrides [103->104 ]
jigasi_1   | Jigasi 2020-06-30 21:14:00.259 INFO: [93] service.protocol.media.CallPeerMediaHandler.start().1961 Starting
jicofo_1   | Jicofo 2020-06-30 21:14:00.260 INFO: [30] org.jitsi.jicofo.JitsiMeetConferenceImpl.log() Got session-accept from: bottomchallengesdistractexactly@muc.meet.jitsi/201de70d
jicofo_1   | Jicofo 2020-06-30 21:14:00.261 INFO: [30] org.jitsi.jicofo.JitsiMeetConferenceImpl.log() Received session-accept from bottomchallengesdistractexactly@muc.meet.jitsi/201de70d with accepted sources:Sources{ audio: [ssrc=811271459 ] }@1781972221
jicofo_1   | Jicofo 2020-06-30 21:14:00.262 INFO: [30] org.jitsi.protocol.xmpp.AbstractOperationSetJingle.log() Notify add SSRC bottomchallengesdistractexactly@muc.meet.jitsi/fe792423 SID: e35qbhia306dk Sources{ audio: [ssrc=811271459 ] }@1366289212 source_Groups{ }@1662135123
jicofo_1   | Jicofo 2020-06-30 21:14:00.262 INFO: [30] org.jitsi.protocol.xmpp.AbstractOperationSetJingle.log() Notify add SSRC bottomchallengesdistractexactly@muc.meet.jitsi/201de70d SID: a1negrpcoceh3 Sources{ audio: [ssrc=2654748191 ] }@1472498715 source_Groups{ }@139068234
web_1      | 109.252.78.223 - - [30/Jun/2020:21:14:00 +0000] "POST /http-bind?room=bottomchallengesdistractexactly HTTP/2.0" 200 1076 "https://kappa.alphacephei.com:8443/BottomChallengesDistractExactly" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36"
jvb_1      | Jun 30, 2020 9:14:00 PM org.jitsi.utils.logging2.LoggerImpl log
jvb_1      | INFO: Assume that the remote side is Jigasi, we'll act as server
jvb_1      | Jun 30, 2020 9:14:00 PM org.jitsi.utils.logging2.LoggerImpl log
jvb_1      | INFO: Negotiated DTLS version DTLS 1.0
jvb_1      | Jun 30, 2020 9:14:00 PM org.jitsi.utils.logging2.LoggerImpl log
jvb_1      | INFO: DTLS handshake complete
jvb_1      | Jun 30, 2020 9:14:00 PM org.jitsi.utils.logging2.LoggerImpl log
jvb_1      | INFO: Transport description:
jvb_1      |  <transport xmlns='urn:xmpp:jingle:transports:ice-udp:1' pwd='p3d99v8v28mael42bpit2a1q0' ufrag='52ifc1ec3h8jb8'><rtcp-mux/><fingerprint xmlns='urn:xmpp:jingle:apps:dtls:0' setup='passive' hash='sha-256'>53:CF:DE:40:50:F4:2C:C5:12:51:A0:AE:81:64:2C:9E:EA:79:A6:D2:18:C1:26:2C:55:ED:9E:41:36:7F:09:9E</fingerprint><candidate component='1' foundation='1' generation='0' id='3db5561a70a0fab802ac0484c' network='0' priority='2130706431' protocol='ssltcp' tcptype='passive' type='host' ip='172.26.0.4' port='4443'/><candidate component='1' foundation='3' generation='0' id='568885aa70a0fab802ac05e01' network='0' priority='2130706431' protocol='udp' type='host' ip='172.26.0.4' port='10000'/><candidate component='1' foundation='2' generation='0' id='7d2c846270a0fab8013a23efb' network='0' priority='1694498815' protocol='ssltcp' tcptype='passive' type='srflx' ip='148.251.246.179' port='4443' rel-addr='172.26.0.4' rel-port='4443'/><candidate component='1' foundation='4' generation='0' id='400c794c70a0fab8013a254b0' network='0' priority='1677724415' protocol='udp' type='srflx' ip='148.251.246.179' port='10000' rel-addr='172.26.0.4' rel-port='10000'/></transport>
jigasi_1   | Jigasi 2020-06-30 21:14:00.381 INFO: [93] org.jitsi.jigasi.JvbConference.callStateChanged().1202 [ctx=15935516371731903558203] JVB conference call IN_PROGRESS.
jigasi_1   | Jigasi 2020-06-30 21:14:00.420 INFO: [98] org.jitsi.srtp.crypto.OpenSslWrapperLoader.log() jitsisrtp successfully loaded
web_1      | 109.252.78.223 - - [30/Jun/2020:21:14:00 +0000] "GET /images/favicon.ico?v=1 HTTP/2.0" 200 4424 "https://kappa.alphacephei.com:8443/BottomChallengesDistractExactly" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36"
web_1      | 109.252.78.223 - - [30/Jun/2020:21:14:00 +0000] "POST /http-bind?room=bottomchallengesdistractexactly HTTP/2.0" 200 906 "https://kappa.alphacephei.com:8443/BottomChallengesDistractExactly" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36"
jvb_1      | Jun 30, 2020 9:14:02 PM org.jitsi.utils.logging2.LoggerImpl log
jvb_1      | INFO: ICE state changed from Completed to Terminated.
jvb_1      | Jun 30, 2020 9:14:02 PM org.jitsi.utils.logging2.LoggerImpl log
jvb_1      | INFO: ICE state changed old=Completed new=Terminated
jigasi_1   | Jigasi 2020-06-30 21:14:02.402 WARNING: [87] org.jitsi.jigasi.transcription.Transcriber.log() bottomchallengesdistractexactly@muc.meet.jitsi: participant with identifier focus left while it did not exist
jigasi_1   | Jigasi 2020-06-30 21:14:02.402 INFO: [87] org.jitsi.jigasi.JvbConference.memberPresenceChanged().1006 [ctx=15935516371731903558203] Member left : OWNER focus@auth.meet.jitsi/focus66394518285504428
jigasi_1   | Jigasi 2020-06-30 21:14:02.403 INFO: [87] org.jitsi.jigasi.JvbConference.memberPresenceChanged().1014 [ctx=15935516371731903558203] Focus left! - stopping
jigasi_1   | Jigasi 2020-06-30 21:14:02.403 SEVERE: [87] org.jitsi.jigasi.JvbConference.stop().498 [ctx=15935516371731903558203] Already stopped !
web_1      | 109.252.78.223 - - [30/Jun/2020:21:14:02 +0000] "POST /http-bind?room=bottomchallengesdistractexactly HTTP/2.0" 200 907 "https://kappa.alphacephei.com:8443/BottomChallengesDistractExactly" "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_12_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.116 Safari/537.36"
jigasi_1   | Jigasi 2020-06-30 21:14:02.406 INFO: [87] org.jitsi.jigasi.transcription.GoogleCloudTranscriptionService.log() bottomchallengesdistractexactly/fe792423: sent 1500ms to speech API, for a total of 1 intervals with a total of 3 requests.
jicofo_1   | Jicofo 2020-06-30 21:14:02.407 INFO: [70] org.jitsi.jicofo.jigasi.TranscriberManager.log() detected transcription status being turned off.
jicofo_1   | Jicofo 2020-06-30 21:14:02.407 INFO: [30] org.jitsi.jicofo.ChatRoomRoleAndPresence.log() Chat room event ChatRoomMemberPresenceChangeEvent[type=MemberLeft sourceRoom=org.jitsi.impl.protocol.xmpp.ChatRoomImpl@29e011d6 member=ChatMember[bottomchallengesdistractexactly@muc.meet.jitsi/201de70d, jid: null]@1190302634]
jicofo_1   | Jicofo 2020-06-30 21:14:02.407 INFO: [30] org.jitsi.jicofo.JitsiMeetConferenceImpl.log() Member bottomchallengesdistractexactly@muc.meet.jitsi/201de70d is leaving
jicofo_1   | Jicofo 2020-06-30 21:14:02.408 INFO: [30] org.jitsi.jicofo.JitsiMeetConferenceImpl.log() Terminating: bottomchallengesdistractexactly@muc.
```

**Comment by nshmyrev** [comment]
And I do not have caption button :(

**Comment by nshmyrev** [comment]
Nevermind, I figured out how to add caption button in interface_config.js. No idea why is it hidden. I will add documentation about it. Transcription works now.

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java (hunk 1 above)
Comment (nshmyrev) [reply src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java]: Fixed

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java (hunk 2 above)
Comment (nshmyrev) [reply src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java]: Fixed

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java (hunk 4 above)
Comment (nshmyrev) [reply src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java]: Fixed

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java (hunk 5 above)
Comment (nshmyrev) [reply src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:22]: Fixed

**Comment by nshmyrev** [comment]
> Do you think it's realistic to run a Jigasi instance and a docker instance running Vosk on the same hardware?

@nikvaessen yes, sure. The modern server with 8 cores can decode 10 parallel streams or so, so if you have less than 10 participants you can perfectly run with transcription.

**Comment by nshmyrev** [comment]
Pull request about missing captions button https://github.com/jitsi/docker-jitsi-meet/pull/665

**Comment by nshmyrev** [comment]
@nikvaessen @damencho please review once you have time

**Comment by damencho** [comment]
Hi, thanks for your contribution!
If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)? We would unfortunately be unable to merge your patch unless we have that piece :(.

File: src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java [hunk 7]
```diff
@@ -0,0 +1,349 @@
+/*
+ * Jigasi, the JItsi GAteway to SIP.
+ *
+ * Copyright @ 2020 Atlassian Pty Ltd
```
Comment (damencho) [inline src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java:4]: Can you fix this line saying 8x8, you will find it in other files in the repo.

Comment (nshmyrev) [reply src/main/java/org/jitsi/jigasi/transcription/VoskTranscriptionService.java]: Fixed

**Comment by nshmyrev** [comment]
> If you haven't already done so, could you please make sure you sign our CLA (https://jitsi.org/icla for individuals and https://jitsi.org/ccla for corporations)?

Done

**Review by damencho — APPROVED** [review]

**Comment by nshmyrev** [comment]
Thank you colleagues for developing jitsi!

**Comment by 88fingerslukee** [comment]
@nshmyrev There's a thread with an issue over here:

https://community.jitsi.org/t/jigasi-transcription-with-vosk-repeated-lines/91573/10

could you kindly please look at it?

**Comment by nshmyrev** [comment]
I will check coming days and let you know

**Pull request merged**